Allow provision RWM volume on the requested zone provided by topology requirements in WCP

akankshapanse opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR includes changes to allow provision RWM volume on the requested zone provided by topology requirements in WCP, when workload domain isolation FSS is enabled.
There are certain changes needed from https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/2985, which will complete the overall changes needed for RWM support in multi-zone cluster

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:
Added unittests to verify the changes, that passed successfully

Testcase in GC controller
```
=== RUN   TestGuestClusterControllerFlowForWorkloadDomainIsolation
{"level":"info","time":"2024-08-31T10:03:39.142563984+05:30","caller":"wcpguest/controller.go:249","msg":"CreateVolume: called with args {Name:pvc-12345 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[access_mode:<mode:MULTI_NODE_MULTI_WRITER > ] Parameters:map[svstorageclass:test-storageclass] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:requisite:<segments:<key:\"R1\" value:\"Zone1\" > > preferred:<segments:<key:\"R1\" value:\"Zone1\" > >  MutableParameters:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"bfa07701-73bf-43e1-8b79-6a967d77473b"}
{"level":"info","time":"2024-08-31T10:03:39.142805301+05:30","caller":"wcpguest/controller_helper.go:309","msg":"Waiting up to 240 seconds for PersistentVolumeClaim -12345 in namespace test-namespace to have phase Bound","TraceId":"bfa07701-73bf-43e1-8b79-6a967d77473b"}
{"level":"info","time":"2024-08-31T10:03:40.14220787+05:30","caller":"wcpguest/controller_helper.go:333","msg":"PersistentVolumeClaim -12345 in namespace test-namespace is in state Bound","TraceId":"bfa07701-73bf-43e1-8b79-6a967d77473b"}
{"level":"info","time":"2024-08-31T10:03:40.142395154+05:30","caller":"wcpguest/controller.go:421","msg":"Volume \"-12345\" created is accessible from zones: [map[R1:Zone1]]","TraceId":"bfa07701-73bf-43e1-8b79-6a967d77473b"}
{"level":"info","time":"2024-08-31T10:03:40.142431515+05:30","caller":"wcpguest/controller.go:446","msg":"Volume created successfully. Volume Handle: \"-12345\", PV Name: \"pvc-12345\"","TraceId":"bfa07701-73bf-43e1-8b79-6a967d77473b"}
    controller_test.go:404: AccessibleTopology was correctly set in create volume response
{"level":"info","time":"2024-08-31T10:03:40.143824679+05:30","caller":"wcpguest/controller.go:465","msg":"DeleteVolume: called with args: {VolumeId:-12345 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"a5483328-2d6e-425c-908b-939801c9623b"}
{"level":"info","time":"2024-08-31T10:03:40.143947333+05:30","caller":"wcpguest/controller.go:510","msg":"DeleteVolume: Volume deleted successfully. VolumeID: \"-12345\"","TraceId":"a5483328-2d6e-425c-908b-939801c9623b"}
{"level":"info","time":"2024-08-31T10:03:40.143971492+05:30","caller":"wcpguest/controller.go:525","msg":"Volume \"-12345\" deleted successfully.","TraceId":"a5483328-2d6e-425c-908b-939801c9623b"}
--- PASS: TestGuestClusterControllerFlowForWorkloadDomainIsolation (2.00s)
```

Testcase in WCP controller - only negative scenarios as unittesting framework needs to be enhanced for file volume support

```
=== RUN   TestWCPCreateVolumeWithoutZoneLabelPresentForFileVolume
{"level":"info","time":"2024-08-31T16:09:52.521845996+05:30","caller":"wcp/controller.go:1135","msg":"CreateVolume: called with args {Name:test-pvc-1451bd23-8c2c-4e3d-9671-61c8f25c0745 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[access_mode:<mode:MULTI_NODE_MULTI_WRITER > ] Parameters:map[storagepolicyid:aa6d5a82-1c88-45da-85d3-3d74b91a5bad] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements: MutableParameters:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"f36c97a4-0e83-4d9f-838c-658b141a8b32"}                                                             {"level":"error","time":"2024-08-31T16:09:52.521877939+05:30","caller":"wcp/controller.go:979","msg":"volume provisioning request received without topologyRequirement.","TraceId":"f36c97a4-0e83-4d9f-838c-658b141a8b32","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).createFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:979\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume.func1\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1179\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1183\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.TestWCPCreateVolumeWithoutZoneLabelPresentForFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller_test.go:404\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1595"}
{"level":"error","time":"2024-08-31T16:09:52.521899509+05:30","caller":"wcp/controller.go:1190","msg":"Operation failed, reporting failure status to Prometheus. Operation Type: \"create-volume\", Volume Type: \"file\", Fault Type: \"csi.fault.Internal\"","TraceId":"f36c97a4-0e83-4d9f-838c-658b141a8b32","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1190\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.TestWCPCreateVolumeWithoutZoneLabelPresentForFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller_test.go:404\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1595"}
    controller_test.go:406: expected error is thrown: rpc error: code = FailedPrecondition desc = volume provisioning request received without topologyRequirement.
--- PASS: TestWCPCreateVolumeWithoutZoneLabelPresentForFileVolume (0.00s)

=== RUN   TestWCPCreateVolumeWithHostLabelPresentForFileVolume
{"level":"info","time":"2024-08-31T16:09:52.526621865+05:30","caller":"wcp/controller.go:1135","msg":"CreateVolume: called with args {Name:test-pvc-8327d65b-6d02-453e-a0d5-1946e8bc4ac3 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[access_mode:<mode:MULTI_NODE_MULTI_WRITER > ] Parameters:map[storagepolicyid:aa6d5a82-1c88-45da-85d3-3d74b91a5bad] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:preferred:<segments:<key:\"kubernetes.io/hostname\" value:\"host1\" > >  MutableParameters:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"9a7a98c8-cdfd-4f00-b18e-52a79c5ecf72"}
{"level":"error","time":"2024-08-31T16:09:52.526706824+05:30","caller":"wcp/controller.go:971","msg":"support for topology requirement with hostname labels is not yet implemented ","TraceId":"9a7a98c8-cdfd-4f00-b18e-52a79c5ecf72","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).createFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:971\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume.func1\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1179\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1183\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.TestWCPCreateVolumeWithHostLabelPresentForFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller_test.go:501\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1595"}
{"level":"error","time":"2024-08-31T16:09:52.526741577+05:30","caller":"wcp/controller.go:1190","msg":"Operation failed, reporting failure status to Prometheus. Operation Type: \"create-volume\", Volume Type: \"file\", Fault Type: \"csi.fault.Internal\"","TraceId":"9a7a98c8-cdfd-4f00-b18e-52a79c5ecf72","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.(*controller).CreateVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller.go:1190\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/wcp.TestWCPCreateVolumeWithHostLabelPresentForFileVolume\n\t/home/pansea/csi-src/vsphere-csi-driver/pkg/csi/service/wcp/controller_test.go:501\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:1595"}
    controller_test.go:503: expected error is thrown: rpc error: code = Unimplemented desc = support for topology requirement with hostname labels is not yet implemented
--- PASS: TestWCPCreateVolumeWithHostLabelPresentForFileVolume (0.00s)
```

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Allow provision RWM volume on the requested zone provided by topology requirements in WCP when multiple zones are present with workload domain isolation
```

shalini-b commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -975,9 +1161,19 @@ func (c *controller) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequ
+			if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.WorkloadDomainIsolationSupported) &&
+				!c.manager.VcenterConfig.FileVolumeActivated {
+				return nil, csifault.CSIUnimplementedFault, logger.LogNewErrorCode(log, codes.Unimplemented,
+					"file volume service is disabled on cluster(s)")
```
I guess file services are enabled on the supervisor level, is this error message appropriate? 
cc: @divyenpatel

akankshapanse replied:
as per https://confluence.eng.vmware.com/pages/viewpage.action?spaceKey=CNAS&title=Continue+Support+RWM+Volumes+after+adding+new+workload+zone#ContinueSupportRWMVolumesafteraddingnewworkloadzone-UIChanges, file volume creation will be enabled on supervisor through the new toggle option i.e. fileVolumeActivated. 
If the message looks misleading, I will reword to "file volume creation is disabled on cluster(s)"

shalini-b replied:
I meant can we instead say `file services are disabled on supervisor cluster`. If we say cluster(s), they might search for that option on individual clusters.

akankshapanse replied:
done, updated log message

shalini-b commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -844,18 +844,26 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolumeRequest) (
```
You will also need to start running CNSFileAccessConfig controller for RWX provisioning to work. That is currently disabled for multi-zone SV.

akankshapanse replied:
done, updated code to enable CNSFileAccessConfig controller and also install required CRDs

shalini-b commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -933,6 +1039,86 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
+			}
+
+			// Calculate accessible topology for the provisioned volume.
+			datastoreAccessibleTopology, err := c.topologyMgr.GetTopologyInfoFromNodes(ctx,
```
We want to add all the zones the datastore has access to in the supervisor as node affinity terms, not only the ones mentioned in the topology requirement.

akankshapanse replied:
This will be taken care by changes in https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/2985 once merged

divyenpatel commented:
Guest Controller is not publishing `resp.Volume.AccessibleTopology` for file volume.
Can we update Guest Controller?

divyenpatel commented:
/ok-to-test

akankshapanse commented:
> Guest Controller is not publishing `resp.Volume.AccessibleTopology` for file volume. Can we update Guest Controller?

done, updated code with changes in guest controller and also added unittest

divyenpatel reviewed (approved):
/approve
/lgtm
